### ISSUE TITLE
Pagination is slow as the collection size grows

### ISSUE THREAD

**Issue by john-dupuy** [ibutsu/ibutsu-server#6]
We've been experiencing slowdown with our pagination (particularly with our ever-growing `results` collection) and it can be explained by the use of two functions:
   1) [`mongo.results.count`](https://github.com/ibutsu/ibutsu-server/blob/c3d1446ac23135023f82e591b72b6bec6cc0919a/backend/ibutsu_server/controllers/result_controller.py#L86)
   2) the use of `skip` in[ `mongo.results.find`](https://github.com/ibutsu/ibutsu-server/blob/c3d1446ac23135023f82e591b72b6bec6cc0919a/backend/ibutsu_server/controllers/result_controller.py#L89)
Neither of these can efficiently make use of the indexes on the collection. Therefore, as the collection size grows, they become extremely inefficient. https://docs.mongodb.com/v3.6/reference/method/cursor.skip/#using-cursor-skip cf. in particular:
```
The cursor.skip() method requires the server to scan from the beginning of the input results set before beginning to return results. As the offset increases, cursor.skip() will become slower.
```
We have been talking about moving to PSQL, and while I agree that we will experience some performance boost, I think we will hit the same issues IF we leave the implementation of pagination as is. Based on the PSQL docs, we'd be implementing pagination with https://www.postgresql.org/docs/current/queries-limit.html, cf. in particular:
```
The rows skipped by an OFFSET clause still have to be computed inside the server; therefore a large OFFSET might be inefficient.
```
- Therefore, I think we need to rethink how we do pagination
- Options for pagination
    1) A last ID approach like the one [here](https://www.codementor.io/@arpitbhayani/fast-and-efficient-pagination-in-mongodb-9095flbqr)
      - this would force us to traverse pages in a linear manner (1->2->3-> etc)
    2) A hard limit approach where if the number of documents is greater than some cutoff (say 1e4) we have a button to load the next 1e4 if the user goes to the last page

(ii) is my preferred option, because I think it is rare when someone actually wants to navigate to arbitrary pages of results. 

In the implementation of (ii) we could introduce some (low) timeout for the `mongo.results.count` query, if it hits the timeout, we limit the documents to some `MAX_DOCUMENTS`. 

In this way we circumvent the inherent slowness of `count` and because `skip` will never have to skip more than `MAX_DOCUMENTS`. For loading the next batch of documents we could implement something akin to (i).

Some information is found in http://www.ovaistariq.net/404/mysql-paginated-displays-how-to-kill-performance-vs-how-to-improve-performance/

**Comment by john-dupuy** [comment]
@rsnyman I think we should explore these options before shifting totally to PSQL. I think any investigation from this will be very beneficial regardless of the DB we end up using. 
If we find that the performance still suffers after this, we can then look into the PSQL conversion.

**Comment by john-dupuy** [comment]
I am going to try to put together a POC PR to see how much performance we can gain through this.

**Comment by john-dupuy** [comment]
After exploring this, I am thinking that a simple approach may be the best approach.

**Comment by john-dupuy** [comment]
For the results collection/table, we decided on a simpler approach implemented in #8. Closing this issue for now, but will reopen if the changes introduced by #8 are inadequate.
